Implementation of Seguro Phase II

ashelkovnykov opened the pull request:
Includes:
- Seguro Phase II
- Unit tests
- Integration tests
- Benchmark suite
- Benchmark results
- makefile
- Fully commented
- Enhanced README

Notes:
A partially-complete version of Seguro Phase II that follows the design outlined in the proposal more closely is available on a different branch in my fork. It is trivial to `git cherry-pick` or `git rebase` the two branches together. However, I think the version included in this PR is the one we should go with.

ashelkovnykov committed:
Fix build

ashelkovnykov committed:
Refactor project layout and add makefile

ashelkovnykov committed:
Remove unused test.c

ashelkovnykov committed:
Fix compiler warnings

ashelkovnykov committed:
Replace #include guards with #pragma once

ashelkovnykov committed:
Remove unused imports

ashelkovnykov committed:
Add prospective license

ashelkovnykov committed:
Update README

ashelkovnykov committed:
Refactor benchmark suite for simple events

- Create a default benchmark suite
- Fix/improve option parsing for benchmark suite
- Change option parsing to only apply to custom benchmarks
- Extend default target of makefile to build benchmark binary
- Add documentation to README for how to run custom benchmarks

ashelkovnykov committed:
Add personal TODO list to repo so others can track progress

ashelkovnykov committed:
Refactor style and documentation

ashelkovnykov committed:
Fix benchmarking bugs and add additional batch sizes to default suite

ashelkovnykov committed:
Add additional timers

ashelkovnykov committed:
Add Event and Fragment structs

ashelkovnykov committed:
Implement event fragmentation

ashelkovnykov committed:
Add documentation and edit TODO list

ashelkovnykov committed:
Add unit tests for event fragmentation

ashelkovnykov committed:
Update benchmark suite and Add FDB cluster troubleshooting to README

ashelkovnykov committed:
Refactor timers for benchmark

ashelkovnykov committed:
Refactor test/benchmark source locations

ashelkovnykov committed:
Add integration tests

ashelkovnykov committed:
Fix incorrect use of fdb_run_network() and simplify timers

ashelkovnykov committed:
Remove async TODOs

All FDB writes are async, and the tests/benchmarks prove that Seguro is
writing transactions correctly. It's not clear how/why to test async
commands until phase 3.

ashelkovnykov committed:
Add code to read events from FDB

ashelkovnykov committed:
Add integration tests for reading events

ashelkovnykov committed:
Remove option to run custom write benchmark

The reasoning for this is that the default benchmarks should provide
adequate information regarding the performance of different event sizes,
batch sizes, and FoundationDB value sizes. So long as the benchmark
results are recorded, it's unlikely they'll need to be re-run.

The benchmarks are only for tuning purposes: the results will yield the
optimal values to set as macros in the code. In addition, the benchmarks
are being used to independently verify information readily available in
the FoundationDB docs.

ashelkovnykov committed:
Add minor improvements to error logging

ashelkovnykov committed:
Refactor write benchmarks for final time

- Fixed race condition caused by deallocation outside of callbacks
- Added loading bar to benchmarks
- Added minor documentation

ashelkovnykov committed:
Test different FDB value sizes

ashelkovnykov committed:
Document idea for removing headers entirely

matthew-levan commented on src/event.c:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  // Header encodes number of ADDITIONAL fragments
+  header_length = build_header(f_event->header, num_fragments - 1);
```
Just for clarity, does this mean that an event with 10 fragments would have "9" encoded in the header?

ashelkovnykov replied:
That's right - so with a FDB value size of `10 KB` for fragments, if we see an event of size `1.28 MB`, that would be 128 fragments, so the header in fragment 1 (indexed as fragment 0 in the FDB key) would encode that there are 127 additional fragments.

matthew-levan reviewed (changes requested):
This is great work, thank you. Will you please update the proposal doc to match the logic in the code? In addition to that, I just left one comment for you in the code.

ashelkovnykov committed:
Updated proposal to match alternate implementation

matthew-levan commented on doc/PROPOSAL.md:
```diff
@@ -0,0 +1,163 @@
+
+|           | encoding                                                                           | format                         |
+|-----------|------------------------------------------------------------------------------------|--------------------------------|
+| **key**   | `0x00` &#x7c; 64-bit event id (big-endian) &#x7c; 32-bit fragment id (big-endian)  | `0x00112233445566778899AABBCC` |
```
What determines whether the event id is 32- or 64-bit?

ashelkovnykov replied:
Are you asking, "why use a 64-bit int for the event id but a 32-bit int for the fragment id"?

matthew-levan replied:
Ah I see, misread. Sorry!

matthew-levan commented on doc/PROPOSAL.md:
```diff
@@ -0,0 +1,163 @@
+|------|------------------------|-----------------|-------------|
+| 0    | *additional* fragments | `<# fragments>` | `0111 1111` |
+
+**129 fragments or less**
```
Shouldn't this be "129 fragments or more"?

ashelkovnykov replied:
Yes, that's a typo

matthew-levan commented on doc/PROPOSAL.md:
```diff
@@ -0,0 +1,163 @@
+- The first fragment (fragment `0`) will have a slightly different encoding for
+  the value, as documented below
+
+#### Fragment 0
```
Will you put the reasoning behind using a different encoding for Fragment 0 here please?

ashelkovnykov replied:
The reasoning is already included, check line 68

matthew-levan commented on doc/PROPOSAL.md:
```diff
@@ -0,0 +1,163 @@
+|------|---------------|----------------------------------------|-------------|
+| 0    | byte-width    | overflow toggle bit, bucket byte-width | `1000 0010` |
+| 1    | `# fragments` | total fragments                        | `0000 0000` |
+| 2    | `# fragments` | total fragments                        | `0000 0001` |
```
If this is supposed to encode the *additional* number of fragments left in the event, and this is an example of the last fragment in the sequence, shouldn't the example data read `0000 0000`?

Also perhaps we should replace "total fragments" with "remaining fragments". I actually like "remaining" better than "additional" because it more strongly implies a dynamic counter. Thoughts?

ashelkovnykov replied:
I think you're misunderstanding. Only the first fragment, i.e. fragment 0, has a header. This header includes the number of additional/remaining fragments. Every other fragment has no header at all.

Re: additional vs. remaining - whichever you prefer

matthew-levan replied:
Ah, got it. Ok it would be great to make that more apparent, with a sentence saying exactly what you said: "Only fragment 0 has a header, the rest have none at all."

matthew-levan commented on doc/PROPOSAL.md:
```diff
@@ -0,0 +1,163 @@
+| 1    | `# fragments` | total fragments                        | `0000 0000` |
+| 2    | `# fragments` | total fragments                        | `0000 0001` |
+
+Remember: the header encodes the number of **additional** fragments.
```
If you agree about using the word *remaining* instead of *additional*, let's find/replace all the occurrences.

matthew-levan commented:
Looks good. Only a few more questions above, then I'll approve and merge. Thank you.

ashelkovnykov commented:
GH isn't showing it well, but I pushed a new version of the final commit. It fixed the typo for "129 fragments or more", fixed inconsistent usage of "additional"/"total" to "remaining", and added a little more explanation to the fragment header.
